File: core/java/src/main/java/hara/lang/protocol/IWorkStore.java
package hara.lang.protocol;

import hara.lang.declaration.HaraAvailability;
import hara.lang.declaration.HaraMethod;
import hara.lang.declaration.HaraProtocolBinding;

/** Queries and atomically journals managed Work execution state. */
@HaraProtocolBinding(
    namespace = "std.protocol.iworkstore",
    name = "IWorkStore",
    availability = HaraAvailability.CAPABILITY_GATED,
    capability = "native-runtime-protocols")
public interface IWorkStore {
  @HaraMethod(value = "work-query", arity = 2)
  Object workQuery(Object query);

  @HaraMethod(value = "work-transact", arity = 2)
  Object workTransact(Object transition);
}

File: core/rust/src/lang/protocol/iworkstore.rs
use hara_protocol_macros::hara_protocol;

#[hara_protocol(
    namespace = "std.protocol.iworkstore",
    name = "IWorkStore",
    availability = "capability-gated",
    capability = "native-runtime-protocols"
)]
pub trait IWorkStore {
    type Query;
    type Transition;
    type Output;

    #[hara_method(value = "work-query", arity = 2)]
    fn work_query(&self, query: Self::Query) -> Self::Output;
    #[hara_method(value = "work-transact", arity = 2)]
    fn work_transact(&self, transition: Self::Transition) -> Self::Output;
}
